assignments/this.js

```diff
@@ -1,26 +1,68 @@
 /* The for principles of "this";
 * in your own words. explain the four principle for the "this" keyword below.
 *
-* 1. 
-* 2. 
-* 3. 
-* 4. 
+* 1. Window (Default) Binding: this binds to the global object. It is the default setting.
+* 2. Implicit Binding: occurs when dot notation is used to invoke a function. 
+Whatever is to the left of the dot becomes the context for ```this``` in the function.
+* 3. Explicit Binding: occurs when .call(), .apply(), or .bind() are used on a function.
+* 4. New Binding: occurs when building new objects or when explicitly binding an existing function to the object 
 *
 * write out a code example of each explanation above
 */
 
 // Principle 1
-
 // code example for Window Binding
+function myWinBindFunc () {
+  console.log(this)
+}
+myWinBindFunc()
 
-// Principle 2
+function myWinBindFuncStrict () {
+  'use strict'
+  console.log(this)
+}
+myWinBindFuncStrict() //undefined
 
+// Principle 2
 // code example for Implicit Binding
+let animal = {
+  species: 'dog',
+  name: 'Snoopy',
+  identify: function () {
+    console.log(`${this.name} is a ${this.species}`)
+  }
+}
+animal.identify()
 
 // Principle 3
-
 // code example for New Binding
+// Use a 'constructor' which is a function that creates an object
+function Person(name) {
+  this.name = name,
+  this.greeting = function () {
+    return `Hello, my name is ${this.name}`
+  }
+}
+
+let wilfred = new Person('Wilfred')
+console.log(wilfred)
+console.log(wilfred.greeting())
 
 // Principle 4
+// code example for Explicit Binding
+
```

Comment:
Probably the best example of explicit binding I've seen. It's difficult for me to come up with a good reason to use it.

Reply:
> It's difficult for me to come up with a good reason to use it.

This particular code allows you to combine an object and an array and dynamically iterate through the array (instead of listing item1, item2, item3, etc).

An example I could think of for using this code in real life for creating a mail merge of customer information into a marketing form letter. So if I take my code from Exercise #4 and extend it for this example we've got something like this....

```
function Customer (name, package) {
  this.name = name,
  this.package = package
  this.greeting = function () {
    return `Hello ${this.name}`
  }
}
//Creating an object from the Customer constructor

const joe = new Customer('Joe', premium) //Premium Service Customer
//Create an array of a list of services
const premiumPackage = ['ESPN', 'Disney Channel', 'HBO', 'Cinemax', 'Showtime', 'Starz']

const wilfred = new Customer('Wilfred', basic) //Basic Service Customer
//Create an array of a list of services
const basicPackage = ['ESPN', 'Disney Channel']

// Automatically generate new customer welcome letter
function welcomeLetter (list) {
  return `${this.greeting()}, we see you've signed up for our ${this.package} channels: ${list.map((item, index) => index !== list.length-1 ?' '+ item: ' and ' + item)}.`
}
// Explicity bind the object and array to the introduce function using the .call() method
console.log(welcomeLetter.call(joe, premiumPackage))
console.log(welcomeLetter.call(wilfred, basicPackage))
`
```

Being able to do things like that will make you a hero with the marketing folks and management. Haha!

There's more you can do with this code but this shows the possible value of using explicit binding.